ros_ws/src/teleoperation/src/keyboard_controller.cpp

```diff
@@ -1,84 +1,149 @@
 #include "keyboard_controller.h"
+using std::abs;
 
+double clamp(double value, double lower, double upper)
+{
+    return std::min(upper, std::max(value, lower));
+}
+
+double map(double value, double in_lower, double in_upper, double out_lower, double out_upper)
+{
+    return out_lower + (out_upper - out_lower) * (value - in_lower) / (in_upper - in_lower);
+}
+
+/**
+ * Class constructor that sets up a publisher for the drive parameters topic, creates a window and starts a timer for
+ * the main loop
+ * */
 KeyboardController::KeyboardController()
 {
-    this->drive_parameters_publisher = this->node_handle.advertise<drive_msgs::drive_param>(TOPIC_DRIVE_PARAMETERS, 1);
+    this->m_drive_parameters_publisher =
+        this->m_node_handle.advertise<drive_msgs::drive_param>(TOPIC_DRIVE_PARAMETERS, 1);
+    this->createWindow();
+
+    auto tick_duration = ros::Duration(1.0 / PARAMETER_UPDATE_FREQUENCY);
+    this->m_timer = this->m_node_handle.createTimer(tick_duration, &KeyboardController::timerCallback, this);
+}
+
+KeyboardController::~KeyboardController()
+{
+    SDL_DestroyWindow(this->m_window);
+    SDL_Quit();
 }
 
-void KeyboardController::keyboardLoop()
+void KeyboardController::createWindow()
 {
-    std::cout << "Listening to keyboard..." << std::endl;
-    std::cout << "========================" << std::endl;
+    std::string icon_filename = ros::package::getPath("teleoperation") + std::string("/wasd.bmp");
 
-    double velocity = 0;
-    double angle = 0;
-    while (ros::ok())
+    if (SDL_Init(SDL_INIT_VIDEO) < 0)
     {
-        auto key = static_cast<Keycode>(this->getKeyboardCharacter());
+        throw std::runtime_error("Could not initialize SDL");
+    }
+    this->m_window = SDL_CreateWindow("Keyboard teleoperation - Use WASD keys", SDL_WINDOWPOS_UNDEFINED,
+                                      SDL_WINDOWPOS_UNDEFINED, 450, 100, SDL_WINDOW_RESIZABLE);
+
+    SDL_Surface* icon = SDL_LoadBMP(icon_filename.c_str());
+    if (icon != NULL)
+    {
+        SDL_SetWindowIcon(this->m_window, icon);
+        SDL_FreeSurface(icon);
+    }
+}
 
-        switch (key)
+void KeyboardController::pollWindowEvents()
+{
+    SDL_Event event;
+    while (SDL_PollEvent(&event))
+    {
+        if (event.type == SDL_KEYUP || event.type == SDL_KEYDOWN)
         {
-            case Keycode::W:
-                velocity += 1;
-                break;
-            case Keycode::S:
-                velocity -= 1;
-                break;
-            case Keycode::A:
-                angle += 1;
-                break;
-            case Keycode::D:
-                angle -= 1;
-                break;
-            // Ignore other Keys
-            default:
-                break;
+            for (size_t i = 0; i < KEY_COUNT; i++)
```

Comment:
You could add an assertion here to make sure, `KEY_CODES` and `m_key_pressed_state` have the same size. Because both are arrays a [static assertion](https://en.cppreference.com/w/cpp/language/static_assert) should be possible e.g.
```C++
static_assert(KEY_CODES.size() == m_key_pressed_state.size(), "KEY_CODES and m_key_pressed_state needs to be of the same size");
```